File: dev/20190415/k2works/fizzbuzz/src/main/java/org/hiroshima_arc/domain/type/IFizzBuzzType.java
package org.hiroshima_arc.domain.type;

import org.hiroshima_arc.domain.model.FizzBuzzValue;

public abstract class IFizzBuzzType {
    public FizzBuzzValue generate(int number) {
        boolean fizz = isFizz(number);
        boolean buzz = isBuzz(number);
        if (fizz && buzz) return  new FizzBuzzValue(number,"FizzBuzz");
        if (fizz) return new FizzBuzzValue(number, "Fizz");
        if (buzz) return new FizzBuzzValue(number,"Buzz");
        return new FizzBuzzValue(number,Integer.toString(number));
    }

    protected boolean isBuzz(int number) {
        return number % 5 == 0;
    }

    protected boolean isFizz(int number) {
        return number % 3 == 0;
    }
}

File: dev/20190822/k2works/src/IFizzBuzzType.ts
import { FizzBuzzValue } from "./FizzBuzzValue";

export abstract class IFizzBuzzType {
  public static readonly FIZZ: string = "Fizz";
  public static readonly BUZZ: string = "Buzz";
  public static readonly FIZZ_BUZZ: string = "FizzBuzz";
  public abstract generate(input: number): FizzBuzzValue;
}
